### PR TITLE
Fixes blockwise

### PR DESCRIPTION
**Pull request by boaks**
Correct several issues in blockwise transfer.

Signed-off-by: Achim Kraus achim.kraus@bosch-si.com

### CODE REVIEW

**Comment by boaks** [comment]
Seems that the order of the commits is mixed up.

File: californium-core/src/main/java/org/eclipse/californium/core/network/stack/BlockwiseLayer.java [hunk 1]
```diff
@@ -556,34 +558,69 @@ public void receiveResponse(final Exchange exchange, final Response response) {
...
+								Level.FINER,
+								"discarding outdated block2 transfer response for old notification {0}, current is {1}",
+								new Object[]{ status.getObserve(), response });
+						clearBlock2Status(key);
+						status.completeTransfer(exchange);
 					} else {
-						resp.setType(Type.NON);
+						LOGGER.log(
+								Level.FINER,
+								"discarding old notification [{0}] received during ongoing blockwise transfer: {1}",
```
Comment (sbernard31) [inline californium-core/src/main/java/org/eclipse/californium/core/network/stack/BlockwiseLayer.java]: The log talk about "notification" but It seems to me that we can have block of "normal" response.

File: californium-core/src/test/java/org/eclipse/californium/core/test/lockstep/ObserveClientSideTest.java [hunk 2]
```diff
@@ -172,9 +198,9 @@ public void testBlockwiseObserve() throws Exception {
 		client.addNotificationListener(notificationListener);
 		client.sendRequest(request);
 
-		server.expectRequest(CON, GET, path).storeBoth("A").storeToken("T").go();
+		server.expectRequest(CON, GET, path).observe(0).storeBoth("A").storeToken("T").go();
 		server.sendResponse(ACK, CONTENT).loadBoth("A").observe(0).block2(0, true, 16).size2(respPayload.length()).payload(respPayload.substring(0, 16)).go();
-		server.expectRequest(CON, GET, path).storeBoth("B").block2(1, false, 16).go();
+		server.expectRequest(CON, GET, path).noOption(OptionNumberRegistry.OBSERVE).storeBoth("B").block2(1, false, 16).go();
```
Comment (boaks) [inline californium-core/src/test/java/org/eclipse/californium/core/test/lockstep/ObserveClientSideTest.java]: I will remove the added checks.

**Comment by boaks** [comment]
> There is a lot of change in c979964 and I'm not sure to understand the ins and outs of this.
> I try to rebase that branch on #373 and the tests I added fail because in this version request is not cancelled anymore. (cancelTransfer replaced by completeTransfer)

The issue with the "cancel" is demonstrated in PR #379. The first "trick", to ignore the "cancel" for "recreated" observe exchanges doesn't work, if the first notification (response to observe) is interrupted by a fresh notification. Therefore I started to cleanup that mechanism.

Maybe adding a "cancel" for "none observe get" will then cancel such a interrupted get.

**Comment by boaks** [comment]
> I try to rebase that branch on #373 and the tests I added fail because in this version request is not cancelled anymore. 

Should work now.

**Comment by sbernard31** [comment]
LGTM : I integrated it in master.

**Pull request merged**